File: Core/State/Impl/RouteStateForLoc.cs
using System.Collections.Generic;
using System.Linq;

namespace BinkyRailways.Core.State.Impl
{
    /// <summary>
    /// State of a single route for a specific loc.
    /// </summary>
    internal sealed class RouteStateForLoc : IRouteStateForLoc
    {
        private readonly ILocState loc;
        private readonly IRouteState route;
        private readonly Dictionary<ISensorState, IRouteEventBehaviorState> behaviors = new Dictionary<ISensorState, IRouteEventBehaviorState>();

        /// <summary>
        /// Default ctor
        /// </summary>
        internal RouteStateForLoc(ILocState loc, IRouteState route)
        {
            this.loc = loc;
            this.route = route;

            foreach (var @event in route.Events)
            {
                var behavior = @event.Behaviors.FirstOrDefault(x => x.AppliesTo(loc));
                if (behavior != null) 
                {
                    behaviors[@event.Sensor] = behavior;
                }
            }
        }

        /// <summary>
        /// Gets the loc for which this route state is
        /// </summary>
        public ILocState Loc { get { return loc; } }

        /// <summary>
        /// Gets the underlying route state
        /// </summary>
        public IRouteState Route { get { return route; } }

        /// <summary>
        /// Try to get the behavior for the given sensor.
        /// </summary>
        public bool TryGetBehavior(ISensorState sensor, out IRouteEventBehaviorState behavior)
        {
            if (behaviors.TryGetValue(sensor, out behavior))
            {
                behaviors.Remove(sensor); // Avoid listening to the sensor twice
                return true;
            }
            return false;
        }

        /// <summary>
        /// Does this route contain an event for the given sensor for my loc?
        /// </summary>
        public bool Contains(ISensorState sensor)
        {
            return behaviors.ContainsKey(sensor);
        }
    }
}

File: pkg/core/state/impl/route_state_for_loc.go
// Copyright 2021 Ewout Prangsma
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
// Author Ewout Prangsma
//

package impl

import (
	"github.com/binkyrailways/BinkyRailways/pkg/core/state"
)

// RouteStateForLoc adds implementation functions to state.RouteStateForLoc.
type RouteStateForLoc interface {
	state.RouteStateForLoc
}

type routeStateForLoc struct {
	loc       state.Loc
	route     Route
	behaviors map[string]state.RouteEventBehavior // sensor.ID -> ...
}

// Create a new entity
func newRouteStateForLoc(loc state.Loc, route Route) RouteStateForLoc {
	rs := &routeStateForLoc{
		loc:       loc,
		route:     route,
		behaviors: make(map[string]state.RouteEventBehavior),
	}
	route.ForEachEvent(func(evt state.RouteEvent) {
		var first state.RouteEventBehavior
		evt.ForEachBehavior(func(b state.RouteEventBehavior) {
			if first != nil {
				return
			}
			if b.AppliesTo(loc) {
				first = b
			}
		})
		if first != nil {
			rs.behaviors[evt.GetSensor().GetID()] = first
		}
	})
	return rs
}

// Gets the loc for which this route state is
func (rs *routeStateForLoc) GetLoc() state.Loc {
	return rs.loc
}

// Gets the underlying route state
func (rs *routeStateForLoc) GetRoute() state.Route {
	return rs.route
}

// Try to get the behavior for the given sensor.
func (rs *routeStateForLoc) TryGetBehavior(sensor state.Sensor) (state.RouteEventBehavior, bool) {
	id := sensor.GetID()
	b, found := rs.behaviors[id]
	return b, found
}

// Does this route contain an event for the given sensor for my loc?
func (rs *routeStateForLoc) Contains(sensor state.Sensor) bool {
	id := sensor.GetID()
	_, found := rs.behaviors[id]
	return found
}
